// send-crypto/build/module/common/apis/blockstream.js
import axios from "axios";
import { sortUTXOs } from "../../lib/utxo";
import { DEFAULT_TIMEOUT } from "./timeout";
const getAPIUrl = (testnet) => `https://blockstream.info/${testnet ? "testnet/" : ""}api`;
const fetchUTXO = (testnet) => async (txHash, vOut) => {
    const apiUrl = getAPIUrl(testnet);
    const utxo = (await axios.get(`${apiUrl}/tx/${txHash}`, {
        timeout: DEFAULT_TIMEOUT,
    })).data;
    const heightResponse = (await axios.get(`${apiUrl}/blocks/tip/height`, {
        timeout: DEFAULT_TIMEOUT,
    })).data;
    const confirmations = utxo.status.confirmed
        ? Math.max(1 + parseInt(heightResponse, 10) - utxo.status.block_height, 0)
        : 0;
    return {
        txHash,
        amount: utxo.vout[vOut].value,
        vOut,
        confirmations,
    };
};
const fetchUTXOs = (testnet) => async (address, confirmations) => {
    const apiUrl = getAPIUrl(testnet);
    const response = await axios.get(`${apiUrl}/address/${address}/utxo`, { timeout: DEFAULT_TIMEOUT });
    const heightResponse = await axios.get(`${apiUrl}/blocks/tip/height`, { timeout: DEFAULT_TIMEOUT });
    return response.data
        .map((utxo) => ({
        txHash: utxo.txid,
        amount: utxo.value,
        vOut: utxo.vout,
        confirmations: utxo.status.confirmed
            ? 1 +
                parseInt(heightResponse.data, 10) -
                utxo.status.block_height
            : 0,
    }))
        .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(sortUTXOs);
};
const fetchTXs = (testnet) => async (address, confirmations = 0, limit = 0) => {
    const apiUrl = getAPIUrl(testnet);
    const response = await axios.get(`${apiUrl}/address/${address}/txs`, { timeout: DEFAULT_TIMEOUT });
    const heightResponse = await axios.get(`${apiUrl}/blocks/tip/height`, { timeout: DEFAULT_TIMEOUT });
    const received = [];
    for (const tx of response.data) {
        for (let i = 0; i < tx.vout.length; i++) {
            const vout = tx.vout[i];
            if (vout.scriptpubkey_address === address) {
                received.push({
                    txHash: tx.txid,
                    amount: vout.value,
                    vOut: i,
                    confirmations: tx.status.confirmed
                        ? 1 +
                            parseInt(heightResponse.data, 10) -
                            tx.status.block_height
                        : 0,
                });
            }
        }
    }
    return received
        .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(sortUTXOs);
};
const broadcastTransaction = (testnet) => async (txHex) => {
    const apiUrl = getAPIUrl(testnet);
    const response = await axios.post(`${apiUrl}/tx`, txHex, {
        timeout: DEFAULT_TIMEOUT,
    });
    return response.data;
};
export const Blockstream = {
    fetchUTXO,
    fetchUTXOs,
    broadcastTransaction,
    fetchTXs,
};
