// send-crypto/build/main/handlers/ERC20/ERC20Handler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC20Handler = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const promiEvent_1 = require("../../lib/promiEvent");
const ethUtils_1 = require("../ETH/ethUtils");
const ERC20ABI_1 = require("./ERC20ABI");
const ERC20s_1 = require("./ERC20s");
const resolveAsset = (network, assetIn) => {
    if (typeof assetIn !== "object") {
        throw new Error("");
    }
    const asset = assetIn;
    if (asset.address) {
        return Object.assign(Object.assign({}, asset), { address: asset.address });
    }
    else {
        const address = (ERC20s_1.ERC20s[network] || {})[asset.name || ""];
        if (!address) {
            throw new Error(`Unknown ERC20 token ${asset.name || JSON.stringify(asset)}`);
        }
        return Object.assign(Object.assign({}, asset), { address });
    }
};
class ERC20Handler {
    constructor(_privateKey, network, _options, sharedState) {
        this._decimals = {};
        // Returns whether or not this can handle the asset
        this.handlesAsset = (asset) => {
            return (typeof asset === "object" &&
                asset.hasOwnProperty("type") &&
                asset.type === "ERC20" &&
                (asset.hasOwnProperty("address") || asset.hasOwnProperty("name")));
        };
        // Address
        this.address = async (asset, options, deferHandler) => deferHandler.address("ETH", options);
        // Balance
        this.getBalance = async (assetIn, options, deferHandler) => {
            const asset = resolveAsset(this.network, assetIn);
            const decimals = await this.decimals(asset);
            return (await this.getBalanceInSats(asset, options, deferHandler)).dividedBy(new bignumber_js_1.default(10).exponentiatedBy(decimals));
        };
        this.getBalanceInSats = async (assetIn, options, deferHandler) => {
            const asset = resolveAsset(this.network, assetIn);
            const address = (options && options.address) ||
                (deferHandler && (await deferHandler.address("ETH", options))) ||
                "";
            return new bignumber_js_1.default(await this.getContract(asset).methods.balanceOf(address).call());
        };
        // Transfer
        this.send = (to, valueIn, assetIn, options, deferHandler) => {
            const asset = resolveAsset(this.network, assetIn);
            const promiEvent = promiEvent_1.newPromiEvent();
            (async () => {
                const contract = this.getContract(asset);
                const method = options.approve
                    ? contract.methods.approve
                    : contract.methods.transfer;
                const call = method(to, valueIn
                    .times(new bignumber_js_1.default(10).exponentiatedBy(await this.decimals(asset)))
                    .toFixed());
                const config = Object.assign({ from: await deferHandler.address("ETH") }, ethUtils_1.getTransactionConfig(options));
                // tslint:disable-next-line: no-object-mutation
                config.gas = await call.estimateGas(config);
                const web3PromiEvent = call.send(config);
                promiEvent_1.forwardEvents(web3PromiEvent, promiEvent);
                web3PromiEvent.then(promiEvent.resolve);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        this.sendSats = (to, valueIn, assetIn, options, deferHandler) => {
            const asset = resolveAsset(this.network, assetIn);
            const promiEvent = promiEvent_1.newPromiEvent();
            (async () => {
                const contract = this.getContract(asset);
                const method = options.approve
                    ? contract.methods.approve
                    : contract.methods.transfer;
                const call = method(to, valueIn.toFixed());
                const config = Object.assign({ from: await deferHandler.address("ETH") }, ethUtils_1.getTransactionConfig(options));
                // tslint:disable-next-line: no-object-mutation
                config.gas = await call.estimateGas(config);
                const web3PromiEvent = call.send(config);
                promiEvent_1.forwardEvents(web3PromiEvent, promiEvent);
                web3PromiEvent.then(promiEvent.resolve);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        this.getContract = (asset) => {
            return new this.sharedState.web3.eth.Contract(ERC20ABI_1.ERC20ABI, resolveAsset(this.network, asset).address);
        };
        this.decimals = async (asset) => {
            const address = resolveAsset(this.network, asset).address;
            if (this._decimals[address]) {
                return this._decimals[address];
            }
            return this.getContract(asset).methods.decimals().call();
        };
        this.network = ethUtils_1.getNetwork(network);
        this.sharedState = sharedState;
    }
}
exports.ERC20Handler = ERC20Handler;
