// @glif/filecoin-wallet-provider/dist/providers/local-node-provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalNodeProvider = void 0;
var filecoin_rpc_client_1 = __importDefault(require("@glif/filecoin-rpc-client"));
var LocalNodeProvider = /** @class */ (function () {
    function LocalNodeProvider(options) {
        this.apiAddress = options.apiAddress;
        this.token = options.token;
        this.jsonRpcEngine = new filecoin_rpc_client_1.default({
            apiAddress: options.apiAddress,
            token: options.token,
        });
    }
    LocalNodeProvider.prototype.newAccount = function () {
        return this.jsonRpcEngine.request('WalletNew', 'secp256k1');
    };
    LocalNodeProvider.prototype.getAccounts = function () {
        return this.jsonRpcEngine.request('WalletList');
    };
    LocalNodeProvider.prototype.sign = function (from, message) {
        return this.jsonRpcEngine.request('WalletSignMessage', from, message);
    };
    return LocalNodeProvider;
}());
exports.LocalNodeProvider = LocalNodeProvider;
