// send-crypto/build/main/common/apis/blockstream.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Blockstream = void 0;
const axios_1 = __importDefault(require("axios"));
const utxo_1 = require("../../lib/utxo");
const timeout_1 = require("./timeout");
const getAPIUrl = (testnet) => `https://blockstream.info/${testnet ? "testnet/" : ""}api`;
const fetchUTXO = (testnet) => async (txHash, vOut) => {
    const apiUrl = getAPIUrl(testnet);
    const utxo = (await axios_1.default.get(`${apiUrl}/tx/${txHash}`, {
        timeout: timeout_1.DEFAULT_TIMEOUT,
    })).data;
    const heightResponse = (await axios_1.default.get(`${apiUrl}/blocks/tip/height`, {
        timeout: timeout_1.DEFAULT_TIMEOUT,
    })).data;
    const confirmations = utxo.status.confirmed
        ? Math.max(1 + parseInt(heightResponse, 10) - utxo.status.block_height, 0)
        : 0;
    return {
        txHash,
        amount: utxo.vout[vOut].value,
        vOut,
        confirmations,
    };
};
const fetchUTXOs = (testnet) => async (address, confirmations) => {
    const apiUrl = getAPIUrl(testnet);
    const response = await axios_1.default.get(`${apiUrl}/address/${address}/utxo`, { timeout: timeout_1.DEFAULT_TIMEOUT });
    const heightResponse = await axios_1.default.get(`${apiUrl}/blocks/tip/height`, { timeout: timeout_1.DEFAULT_TIMEOUT });
    return response.data
        .map((utxo) => ({
        txHash: utxo.txid,
        amount: utxo.value,
        vOut: utxo.vout,
        confirmations: utxo.status.confirmed
            ? 1 +
                parseInt(heightResponse.data, 10) -
                utxo.status.block_height
            : 0,
    }))
        .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(utxo_1.sortUTXOs);
};
const fetchTXs = (testnet) => async (address, confirmations = 0, limit = 0) => {
    const apiUrl = getAPIUrl(testnet);
    const response = await axios_1.default.get(`${apiUrl}/address/${address}/txs`, { timeout: timeout_1.DEFAULT_TIMEOUT });
    const heightResponse = await axios_1.default.get(`${apiUrl}/blocks/tip/height`, { timeout: timeout_1.DEFAULT_TIMEOUT });
    const received = [];
    for (const tx of response.data) {
        for (let i = 0; i < tx.vout.length; i++) {
            const vout = tx.vout[i];
            if (vout.scriptpubkey_address === address) {
                received.push({
                    txHash: tx.txid,
                    amount: vout.value,
                    vOut: i,
                    confirmations: tx.status.confirmed
                        ? 1 +
                            parseInt(heightResponse.data, 10) -
                            tx.status.block_height
                        : 0,
                });
            }
        }
    }
    return received
        .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(utxo_1.sortUTXOs);
};
const broadcastTransaction = (testnet) => async (txHex) => {
    const apiUrl = getAPIUrl(testnet);
    const response = await axios_1.default.post(`${apiUrl}/tx`, txHex, {
        timeout: timeout_1.DEFAULT_TIMEOUT,
    });
    return response.data;
};
exports.Blockstream = {
    fetchUTXO,
    fetchUTXOs,
    broadcastTransaction,
    fetchTXs,
};
