// send-crypto/build/main/handlers/FIL/FILHandler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FILHandler = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const filecoin_message_1 = require("@glif/filecoin-message");
const filecoin_wallet_provider_1 = __importStar(require("@glif/filecoin-wallet-provider"));
const promiEvent_1 = require("../../lib/promiEvent");
const retry_1 = require("../../lib/retry");
const provider_1 = require("./provider");
class FILHandler {
    constructor(privateKey, network, options = {}, sharedState) {
        this.decimals = 18;
        // Returns whether or not this can handle the asset
        this.handlesAsset = (asset) => typeof asset === "string" &&
            ["FIL", "FILECOIN"].indexOf(asset.toUpperCase()) !== -1;
        this.address = async (asset, _options = {}) => {
            if (!this.handlesAsset(asset)) {
                throw new Error(`Asset ${asset} not supported.`);
            }
            const accounts = await this.sharedState.filecoin.wallet.getAccounts(0, 1, this.network);
            return accounts[0];
        };
        // Balance
        this.getBalance = async (asset, options = {}) => new bignumber_js_1.default(await this.sharedState.filecoin.getBalance((options && options.address) || (await this.address(asset))));
        this.getBalanceInSats = async (asset, options = {}) => (await this.getBalance(asset, options)).multipliedBy(new bignumber_js_1.default(10).exponentiatedBy(this.decimals));
        // Transfer
        this.send = (to, value, asset, options = {}) => this.sendSats(to, value.times(new bignumber_js_1.default(10).exponentiatedBy(this.decimals)), asset, options);
        this.sendSats = (to, valueIn, asset, options = {}) => {
            const promiEvent = promiEvent_1.newPromiEvent();
            (async () => {
                const address = await this.address(asset);
                const nonce = await this.sharedState.filecoin.getNonce(address);
                let value = valueIn;
                let message = new filecoin_message_1.Message(Object.assign({ to, from: address, value: valueIn, method: 0, gasFeeCap: 0, gasLimit: 0, gasPremium: 0, nonce, params: undefined }, options));
                message = await this.sharedState.filecoin.gasEstimateMessageGas(message.toLotusType());
                if (options.subtractFee) {
                    const { maxFee: fee, } = await this.sharedState.filecoin.gasEstimateMaxFee(message.toLotusType());
                    if (fee.gt(value)) {
                        throw new Error(`Unable to include fee in value, fee exceeds value (${fee.toFixed()} > ${value.toFixed()})`);
                    }
                    value = value.minus(fee);
                }
                const signed = await this.sharedState.filecoin.wallet.sign(address, message.toLotusType());
                const tx = await this.sharedState.filecoin.sendMessage(signed.Message, signed.Signature.Data);
                const txHash = tx["/"];
                promiEvent.emit("transactionHash", txHash);
                while (true) {
                    const result = await this.sharedState.filecoin.jsonRpcEngine.request("StateSearchMsg", tx);
                    /*
                    {
                        Message: {
                            '/': 'bafy2bzacedtcalpfb5dj3de6jb5lfncqtxw4kia4xotpg3on3p7gcesiwlube'
                        },
                        Receipt: { ExitCode: 0, Return: null, GasUsed: 433268 },
                        ReturnDec: null,
                        TipSet: [
                            {
                            '/': 'bafy2bzacedsqymqnjl2oir7q2rgbwyqscihvebkvr7hfzdq5b2tvcct3cwczy'
                            },
                            {
                            '/': 'bafy2bzacednrjqbsyooob3f76wwct4iupauehvq6wnzrbf6uc4dskptwu4er6'
                            },
                            {
                            '/': 'bafy2bzaceany2in5cadbyf27mecnmzc6ustqfewd5dobxquidzbn3wtnjd52y'
                            }
                        ],
                        Height: 53025
                    }
                    */
                    if (result) {
                        promiEvent.emit("confirmation", 1);
                        break;
                    }
                    await retry_1.sleep(10 * 1000);
                }
                promiEvent.resolve(txHash);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        const filecoin = new filecoin_wallet_provider_1.default(provider_1.SingleKeyProvider(privateKey), {
            apiAddress: options && options.apiAddress,
            token: options && options.token,
        });
        this.network =
            network === "mainnet" ? filecoin_wallet_provider_1.Network.MAIN : filecoin_wallet_provider_1.Network.TEST;
        sharedState.filecoin = filecoin;
        this.sharedState = sharedState;
    }
}
exports.FILHandler = FILHandler;
