// send-crypto/build/main/handlers/FIL/provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SingleKeyProvider = void 0;
// tslint:disable no-submodule-imports
const js_1 = __importDefault(require("@zondax/filecoin-signing-tools/js"));
const filecoin_wallet_provider_1 = require("@glif/filecoin-wallet-provider");
exports.SingleKeyProvider = (privateKey) => {
    // here we close over the private variables, so they aren't accessible to the outside world
    const PRIVATE_KEY = privateKey;
    const { private_hexstring } = js_1.default.keyRecover(PRIVATE_KEY);
    return {
        getAccounts: async (_nStart, _nEnd, network = filecoin_wallet_provider_1.Network.MAIN) => {
            return [
                js_1.default.keyRecover(PRIVATE_KEY, network === filecoin_wallet_provider_1.Network.TEST).address,
            ];
        },
        sign: async (_from, filecoinMessage) => {
            const { signature } = js_1.default.transactionSign({
                to: filecoinMessage.To,
                from: filecoinMessage.From,
                nonce: filecoinMessage.Nonce,
                value: filecoinMessage.Value,
                gaslimit: filecoinMessage.GasLimit,
                gasfeecap: filecoinMessage.GasFeeCap,
                gaspremium: filecoinMessage.GasPremium,
                method: filecoinMessage.Method,
                params: filecoinMessage.Params || "",
            }, Buffer.from(private_hexstring, "hex").toString("base64"));
            return {
                Message: filecoinMessage,
                Signature: {
                    Type: 0,
                    Data: signature.data,
                },
            };
        },
    };
};
