// send-crypto/build/main/handlers/ETH/ETHHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ETHHandler = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const promiEvent_1 = require("../../lib/promiEvent");
const ethUtils_1 = require("./ethUtils");
class ETHHandler {
    constructor(privateKey, network, options, sharedState) {
        this.decimals = 18;
        // Returns whether or not this can handle the asset
        this.handlesAsset = (asset) => typeof asset === "string" &&
            ["ETH", "ETHER", "ETHEREUM"].indexOf(asset.toUpperCase()) !== -1;
        this.address = async (asset, options) => this.unlockedAddress;
        // (await this.sharedState.web3.eth.getAccounts())[0];
        // Balance
        this.getBalance = async (asset, options) => (await this.getBalanceInSats(asset, options)).dividedBy(new bignumber_js_1.default(10).exponentiatedBy(this.decimals));
        this.getBalanceInSats = async (asset, options) => {
            let atBlock;
            if (options && options.confirmations && options.confirmations > 0) {
                const currentBlock = new bignumber_js_1.default(await this.sharedState.web3.eth.getBlockNumber());
                atBlock = currentBlock
                    .minus(options.confirmations)
                    .plus(1)
                    .toNumber();
            }
            const address = (options && options.address) || (await this.address(asset));
            return new bignumber_js_1.default(await this.sharedState.web3.eth.getBalance(address, atBlock));
        };
        // Transfer
        this.send = (to, value, asset, options) => this.sendSats(to, value.times(new bignumber_js_1.default(10).exponentiatedBy(this.decimals)), asset, options);
        this.sendSats = (to, valueIn, asset, optionsIn) => {
            const promiEvent = promiEvent_1.newPromiEvent();
            (async () => {
                const options = optionsIn || {};
                let value = valueIn;
                const txOptions = ethUtils_1.getTransactionConfig(options);
                if (options.subtractFee) {
                    const gasPrice = txOptions.gasPrice ||
                        (await this.sharedState.web3.eth.getGasPrice());
                    const gasLimit = txOptions.gas || 21000;
                    const fee = new bignumber_js_1.default(gasPrice.toString()).times(gasLimit);
                    if (fee.gt(value)) {
                        throw new Error(`Unable to include fee in value, fee exceeds value (${fee.toFixed()} > ${value.toFixed()})`);
                    }
                    value = value.minus(fee);
                }
                const web3PromiEvent = this.sharedState.web3.eth.sendTransaction(Object.assign(Object.assign({ from: await this.address(asset), gas: 21000 }, txOptions), { to, value: value.toFixed() }));
                promiEvent_1.forwardEvents(web3PromiEvent, promiEvent);
                web3PromiEvent.then(promiEvent.resolve);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        this.network = ethUtils_1.getNetwork(network);
        this.privateKey = privateKey;
        const [web3, address] = ethUtils_1.getWeb3(this.privateKey, ethUtils_1.getEndpoint(this.network, options && options.ethereumNode, options && options.infuraKey));
        this.unlockedAddress = address;
        sharedState.web3 = web3;
        this.sharedState = sharedState;
    }
}
exports.ETHHandler = ETHHandler;
