// send-crypto/build/module/handlers/ETH/ETHHandler.js
import BigNumber from "bignumber.js";
import { forwardEvents, newPromiEvent } from "../../lib/promiEvent";
import { getEndpoint, getNetwork, getTransactionConfig, getWeb3, } from "./ethUtils";
export class ETHHandler {
    constructor(privateKey, network, options, sharedState) {
        this.decimals = 18;
        // Returns whether or not this can handle the asset
        this.handlesAsset = (asset) => typeof asset === "string" &&
            ["ETH", "ETHER", "ETHEREUM"].indexOf(asset.toUpperCase()) !== -1;
        this.address = async (asset, options) => this.unlockedAddress;
        // (await this.sharedState.web3.eth.getAccounts())[0];
        // Balance
        this.getBalance = async (asset, options) => (await this.getBalanceInSats(asset, options)).dividedBy(new BigNumber(10).exponentiatedBy(this.decimals));
        this.getBalanceInSats = async (asset, options) => {
            let atBlock;
            if (options && options.confirmations && options.confirmations > 0) {
                const currentBlock = new BigNumber(await this.sharedState.web3.eth.getBlockNumber());
                atBlock = currentBlock
                    .minus(options.confirmations)
                    .plus(1)
                    .toNumber();
            }
            const address = (options && options.address) || (await this.address(asset));
            return new BigNumber(await this.sharedState.web3.eth.getBalance(address, atBlock));
        };
        // Transfer
        this.send = (to, value, asset, options) => this.sendSats(to, value.times(new BigNumber(10).exponentiatedBy(this.decimals)), asset, options);
        this.sendSats = (to, valueIn, asset, optionsIn) => {
            const promiEvent = newPromiEvent();
            (async () => {
                const options = optionsIn || {};
                let value = valueIn;
                const txOptions = getTransactionConfig(options);
                if (options.subtractFee) {
                    const gasPrice = txOptions.gasPrice ||
                        (await this.sharedState.web3.eth.getGasPrice());
                    const gasLimit = txOptions.gas || 21000;
                    const fee = new BigNumber(gasPrice.toString()).times(gasLimit);
                    if (fee.gt(value)) {
                        throw new Error(`Unable to include fee in value, fee exceeds value (${fee.toFixed()} > ${value.toFixed()})`);
                    }
                    value = value.minus(fee);
                }
                const web3PromiEvent = this.sharedState.web3.eth.sendTransaction({
                    from: await this.address(asset),
                    gas: 21000,
                    ...txOptions,
                    to,
                    value: value.toFixed(),
                });
                forwardEvents(web3PromiEvent, promiEvent);
                web3PromiEvent.then(promiEvent.resolve);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        this.network = getNetwork(network);
        this.privateKey = privateKey;
        const [web3, address] = getWeb3(this.privateKey, getEndpoint(this.network, options && options.ethereumNode, options && options.infuraKey));
        this.unlockedAddress = address;
        sharedState.web3 = web3;
        this.sharedState = sharedState;
    }
}
