// send-crypto/build/module/handlers/ERC20/ERC20Handler.js
import BigNumber from "bignumber.js";
import { forwardEvents, newPromiEvent } from "../../lib/promiEvent";
import { getNetwork, getTransactionConfig } from "../ETH/ethUtils";
import { ERC20ABI } from "./ERC20ABI";
import { ERC20s } from "./ERC20s";
const resolveAsset = (network, assetIn) => {
    if (typeof assetIn !== "object") {
        throw new Error("");
    }
    const asset = assetIn;
    if (asset.address) {
        return { ...asset, address: asset.address };
    }
    else {
        const address = (ERC20s[network] || {})[asset.name || ""];
        if (!address) {
            throw new Error(`Unknown ERC20 token ${asset.name || JSON.stringify(asset)}`);
        }
        return { ...asset, address };
    }
};
export class ERC20Handler {
    constructor(_privateKey, network, _options, sharedState) {
        this._decimals = {};
        // Returns whether or not this can handle the asset
        this.handlesAsset = (asset) => {
            return (typeof asset === "object" &&
                asset.hasOwnProperty("type") &&
                asset.type === "ERC20" &&
                (asset.hasOwnProperty("address") || asset.hasOwnProperty("name")));
        };
        // Address
        this.address = async (asset, options, deferHandler) => deferHandler.address("ETH", options);
        // Balance
        this.getBalance = async (assetIn, options, deferHandler) => {
            const asset = resolveAsset(this.network, assetIn);
            const decimals = await this.decimals(asset);
            return (await this.getBalanceInSats(asset, options, deferHandler)).dividedBy(new BigNumber(10).exponentiatedBy(decimals));
        };
        this.getBalanceInSats = async (assetIn, options, deferHandler) => {
            const asset = resolveAsset(this.network, assetIn);
            const address = (options && options.address) ||
                (deferHandler && (await deferHandler.address("ETH", options))) ||
                "";
            return new BigNumber(await this.getContract(asset).methods.balanceOf(address).call());
        };
        // Transfer
        this.send = (to, valueIn, assetIn, options, deferHandler) => {
            const asset = resolveAsset(this.network, assetIn);
            const promiEvent = newPromiEvent();
            (async () => {
                const contract = this.getContract(asset);
                const method = options.approve
                    ? contract.methods.approve
                    : contract.methods.transfer;
                const call = method(to, valueIn
                    .times(new BigNumber(10).exponentiatedBy(await this.decimals(asset)))
                    .toFixed());
                const config = {
                    from: await deferHandler.address("ETH"),
                    ...getTransactionConfig(options),
                };
                // tslint:disable-next-line: no-object-mutation
                config.gas = await call.estimateGas(config);
                const web3PromiEvent = call.send(config);
                forwardEvents(web3PromiEvent, promiEvent);
                web3PromiEvent.then(promiEvent.resolve);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        this.sendSats = (to, valueIn, assetIn, options, deferHandler) => {
            const asset = resolveAsset(this.network, assetIn);
            const promiEvent = newPromiEvent();
            (async () => {
                const contract = this.getContract(asset);
                const method = options.approve
                    ? contract.methods.approve
                    : contract.methods.transfer;
                const call = method(to, valueIn.toFixed());
                const config = {
                    from: await deferHandler.address("ETH"),
                    ...getTransactionConfig(options),
                };
                // tslint:disable-next-line: no-object-mutation
                config.gas = await call.estimateGas(config);
                const web3PromiEvent = call.send(config);
                forwardEvents(web3PromiEvent, promiEvent);
                web3PromiEvent.then(promiEvent.resolve);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        this.getContract = (asset) => {
            return new this.sharedState.web3.eth.Contract(ERC20ABI, resolveAsset(this.network, asset).address);
        };
        this.decimals = async (asset) => {
            const address = resolveAsset(this.network, asset).address;
            if (this._decimals[address]) {
                return this._decimals[address];
            }
            return this.getContract(asset).methods.decimals().call();
        };
        this.network = getNetwork(network);
        this.sharedState = sharedState;
    }
}
