// send-crypto/build/module/handlers/FIL/provider.js
// tslint:disable no-submodule-imports
import filecoin_signer from "@zondax/filecoin-signing-tools/js";
import { Network as FilNetwork, } from "@glif/filecoin-wallet-provider";
export const SingleKeyProvider = (privateKey) => {
    // here we close over the private variables, so they aren't accessible to the outside world
    const PRIVATE_KEY = privateKey;
    const { private_hexstring } = filecoin_signer.keyRecover(PRIVATE_KEY);
    return {
        getAccounts: async (_nStart, _nEnd, network = FilNetwork.MAIN) => {
            return [
                filecoin_signer.keyRecover(PRIVATE_KEY, network === FilNetwork.TEST).address,
            ];
        },
        sign: async (_from, filecoinMessage) => {
            const { signature } = filecoin_signer.transactionSign({
                to: filecoinMessage.To,
                from: filecoinMessage.From,
                nonce: filecoinMessage.Nonce,
                value: filecoinMessage.Value,
                gaslimit: filecoinMessage.GasLimit,
                gasfeecap: filecoinMessage.GasFeeCap,
                gaspremium: filecoinMessage.GasPremium,
                method: filecoinMessage.Method,
                params: filecoinMessage.Params || "",
            }, Buffer.from(private_hexstring, "hex").toString("base64"));
            return {
                Message: filecoinMessage,
                Signature: {
                    Type: 0,
                    Data: signature.data,
                },
            };
        },
    };
};
